stl/inc/syncstream

```diff
@@ -0,0 +1,312 @@
+    basic_syncbuf& operator=(basic_syncbuf&& _Right) {
+        _Emit();
+        _Move_assign(_STD move(_Right), _Choose_pocma<_Alloc>{});
+        return *this;
+    }
+
+    void swap(basic_syncbuf& _Right) {
+        if (this != _STD addressof(_Right)) {
+            _Pocs(_Getal(), _Right._Getal());
+            _Swap_except_al(_Right);
+        }
+    }
+
+    bool emit() {
+        if (_Wrapped == nullptr) {
+            return false;
+        }
+
+        bool _Result                 = true;
+        const size_type _Data_size   = _Get_data_size();
+        const pointer _Begin_seq_ptr = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        if (_Data_size > 0 || _Record_sync) {
+            scoped_lock _Guard(_Get_mutex_for_instance(static_cast<void*>(_Wrapped)));
+
+            if (_Data_size > 0 && _Data_size != _Wrapped->sputn(_Begin_seq_ptr, _Data_size)) {
+                _Result = false;
+            }
+
+            if (_Record_sync) {
+                if (_Wrapped->pubsync() == -1) {
+                    _Result = false;
+                }
+            }
+        }
+        _Record_sync = false;
+        streambuf_type::setp(
+            _Begin_seq_ptr, pointer_traits<pointer>::pointer_to(*streambuf_type::epptr())); // reset written data
+        return _Result;
+    }
+
+    streambuf_type* get_wrapped() const noexcept {
+        return _Wrapped;
+    }
+
+    allocator_type get_allocator() const noexcept {
+        return _Getal();
+    }
+
+    void set_emit_on_sync(bool _Val) noexcept {
+        _Emit_on_sync = _Val;
+    }
+
+protected:
+    int sync() override {
+        _Record_sync = true;
+
+        if (_Emit_on_sync) {
+            if (!emit()) {
+                return -1;
+            }
+        }
+        return 0;
+    }
+
+    int_type overflow(int_type _Current_elem) override {
+        if (_Wrapped == nullptr)
+            return _Traits::eof();
+        const bool _Chk_eof = _Traits::eq_int_type(_Current_elem, _Traits::eof());
+        if (_Chk_eof)
+            return _Traits::not_eof(_Current_elem);
+
+        const size_type _Buf_size       = _Get_buffer_size();
+        const size_type _Max_allocation = allocator_traits<_Alloc>::max_size(_Al);
+        if (_Buf_size == _Max_allocation)
+            return _Traits::eof();
+
+        const size_type _New_capacity  = _Calculate_growth(_Buf_size, _Buf_size + 1, _Max_allocation);
+        const pointer _Old_ptr         = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        const size_type _Old_data_size = _Get_data_size();
+
+        const pointer _New_ptr = _Unfancy(_Al.allocate(_New_capacity));
+        _Traits::copy(_New_ptr, _Old_ptr, _Old_data_size);
+
+        streambuf_type::setp(_New_ptr, _New_ptr + _Old_data_size, _New_ptr + _New_capacity);
+        streambuf_type::sputc(_Current_elem);
+
+        return _Current_elem;
+    }
+
+private:
+    enum { // constant for minimum buffer size
+        _MINSIZE = 32
+    };

```

Comment:
In new code, we prefer to use `static constexpr` data members for such constants, and give them `_Ordinary_ugly_names` instead of `_MACRO_UGLY_NAMES`.